[AMDGPU] Update hasUnwantedEffectsWhenEXECEmpty

perlfu opened the pull request:
Add barriers and s_wait_event to hasUnwantedEffectsWhenEXECEmpty.
Add a comment documenting the current expected use of the function.

perlfu committed:
[AMDGPU] Update hasUnwantedEffectsWhenEXECEmpty

Add barriers and s_wait_event to hasUnwantedEffectsWhenEXECEmpty.
Add a comment documenting the current expected use of the function.

perlfu commented:
Currently includes #97676 as this builts on top of it.

jayfoad reviewed (commented):
Seems reasonable, but I don't really feel super confident about whether this could break anything or fix anything. Does it have any effect on Vulkan CTS?

perlfu commented:
> Seems reasonable, but I don't really feel super confident about whether this could break anything or fix anything. Does it have any effect on Vulkan CTS?

This change passes Vulkan CTS.

Let me clarify why I think this is "correct".
Current codegen can generate branches over barriers, these will be always be retained if we run with -O0.
SIPreEmitPeephole can remove these if the branch distances is short and the function modified in this change returns false.
This leads to the situation where we *might* not skip a barrier for arbitrary reasons.
In practice, if all invocations in the workgroup are running the same shader/kernel in non-cyclic control flow then unskipping the barrier will not make any difference.
However if we have some kind cyclic barrier invocation with divergence, or interaction of different kernels on a barrier, then we should respect the branches over barrier operations, because we do not know any better.

We might of course decide that certain barriers should *always* be executed.
However I think this is for the control flow lowering to decide.
If we want this then we simply need to not introduce a branch over the barrier in the first place.

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4118,6 +4118,13 @@ bool SIInstrInfo::modifiesModeRegister(const MachineInstr &MI) {
 }
 
 bool SIInstrInfo::hasUnwantedEffectsWhenEXECEmpty(const MachineInstr &MI) const {
+  // This function is used to determine if an instruction can be safely
+  // executed under EXECZ without hardware error, indeterminate results,
+  // and/or visible effects on future vector execution or outside the shader.
+  // Note: as of 2024 the only use of this is SIPreEmitPeephole where it is
+  // used in removing branches over short EXECZ sequences.
+  // As such it embeds certain assumptions which may not apply in every case
+  // of EXECZ execution.
```
This should be the doc string in the header

perlfu replied:
Done.

arsenm replied:
You duplicated it, only keep the header one

perlfu replied:
Thanks, that wasn't my intention.

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.h:
```diff
@@ -936,6 +936,14 @@ class SIInstrInfo final : public AMDGPUGenInstrInfo {
            Opcode == AMDGPU::S_BARRIER_SIGNAL_ISFIRST_IMM;
   }
 
+  bool isBarrierRelated(unsigned Opcode) const {
```
Name could be better. Not sure what "barrier related" means, or why DS_GWS_BARRIER doesn't count as a barrier

perlfu replied:
Change to `isBarrier` add DS_GWS parts.

perlfu committed:
Address reviewer feedback:
- Rename isBarrierRelated -> isBarrier
- Add DS_GWS instructions to isBarrier
- Move comment to header

perlfu committed:
- Remove duplicated comment.
